Switch to C++17

jbytheway opened the pull request:
#### Summary
Infrastructure "Convert build to C++17 (from C++14)"

#### Purpose of change
Following up on #63970 and #64011.

We want to move to C++17 to get new features and generally keep the codebase up to date.

#### Describe the solution
Update each of the build files to use C++17 (`Makefile`, `CMakeLists.txt`, Visual Studio projects).

Fix the various compiler errors and clang-tidy warnings that arise.

#### Describe alternatives you've considered
~Waiting until #64011 was already merged.~  It is merged now.

#### Testing
Unit tests, clang-tidy.

Mostly this needs to be tested on CI, so I have to open it as a non-draft, but marking it WIP because there's bound to be issues on other compilers.

#### Additional context

jbytheway committed:
Switch build configs to C++17

Convert the Makefile, CMakeLists, and Visual Studio projects to use
C++17.

jbytheway committed:
C++17 fixes

- std::uncaught_exception is deprecated.
- std::is_literal is deprecated.
- Some small things in the filesystem library have changed.
- [[nodiscard]] is now available, so suppress the clang-tidy check
  suggesting it.

jbytheway committed:
Apply clang-tidy fixes for C++17

A few small changes in C++17 mean that clang-tidy is now able to apply a
few more fixes.  Do so.

jbytheway committed:
Fix clang-tidy-plugin for C++17

irwiss commented:
Considering not all cpp17 features can be used; can structured binding be used? Asking both from compiler support and code style perspectives? E.g.
```
nutrients this_min;
nutrients this_max;
std::tie( this_min, this_max ) = compute_nutrient_range( result_it, rec, extra_flags );
```
into
`auto [this_min, this_max] = compute_nutrient_range(...`
(as you can see type information isn't visible)
Decomposing pairs is by far the most annoying part of using them

jbytheway commented:
From a compiler support perspective, yes, structured bindings are fine.

From a code style perspective, I think we will probably allow them in certain situations.  But since we do have the "almost always avoid `auto`" rule, we'll probably want to be cautious about how we adopt them.  Only time (and @kevingranade) will tell.

At the least I hope to use them for iterating over containers of pairs, which is a super common pattern in our codebase.

jbytheway commented:
For reference, the C++17 features which we should avoid are documented [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/COMPILING/COMPILER_SUPPORT.md),

jbytheway committed:
Improve error message regarding user dir

To help me debug CI issues.

BrettDong commented:
Maybe revert back to ghc file system and see if permission errors go away?

jbytheway commented:
> Maybe revert back to ghc file system and see if permission errors go away?

@akrieger is hoping to test locally and see what can be done.

kevingranade commented:
I saw an instance of this in an unrelated PR yesterday and it resolved with a rerun, so at a minimum I don't think this PR is the cause.

jbytheway commented:
Good to know, thanks!

RAldrich commented:
I would appreciate it if we could resolve #64210 before merging this. (Not going to fix / dropping Debian support is totally a possible resolution - I just think it would be best to make a decision before merging additional compiler changes).

jbytheway commented:
I think the issue in #64210 should be mostly orthogonal to this, but thanks for the heads up.

BrettDong commented:
Successfully reproduced the error. Built with VC2019 build tools.

![image](https://user-images.githubusercontent.com/21075502/224756204-3a33f6c8-4290-4c1a-b888-3be2c3862a6d.png)

BrettDong commented:
The error is caused by the trailing `/` in the path requested. I made it working in VC2019 by removing trailing slash:

![image](https://user-images.githubusercontent.com/21075502/224760106-425dc3a9-0b9b-448b-a4be-a4cb8f2e5045.png)

```diff
diff --git a/src/filesystem.cpp b/src/filesystem.cpp
index e498da3736..d437f50470 100644
--- a/src/filesystem.cpp
+++ b/src/filesystem.cpp
@@ -21,7 +21,11 @@

 bool assure_dir_exist( const std::string &path )
 {
-    return assure_dir_exist( fs::u8path( path ) );
+    std::string p = path;
+    if (!p.empty() && p.back() == '/') {
+        p.pop_back();
+    }
+    return assure_dir_exist( fs::u8path( p ));
 }

 bool assure_dir_exist( const cata_path &path )
```

akrieger commented:
I saw something related to that in the some 2019 patch notes. I believe there are many fixed listed in https://learn.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements-2019?view=msvc-170 which are backports of recent standards committee meetings, including stuff about trailing slashes like mentioned in https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87846

akrieger commented:
Specifically https://cplusplus.github.io/LWG/issue3096 is probably the relevant change.

jbytheway committed:
Work around VS2019 filesystem issue

BrettDong commented on src/filesystem.cpp:
```diff
@@ -21,6 +21,15 @@
 
 bool assure_dir_exist( const std::string &path )
 {
+    // TEMPORARY until we drop VS2019 support
```
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87846 says GCC was behaving the same way until the fix in GCC 8.3. Our documentation says minimum supported GCC is 8.1, so a bump in GCC support policy is also needed to remove this workaround in the future.

jbytheway replied:
That's a doc mistake.  It's supposed to be minimum gcc 9.1 now since dropping Code::Blocks.  I'll PR a fix to the docs.

kevingranade commented:
Regarding the item density test, I thought it was something like that too, but it turns out it's much more prosaic, the clang++-12 workflow is the first one that runs with magiclysm loaded, and you were hitting a magiclysm-specific data error that was fixed in #64258
Other than that, it looks like this will be good to go with a rebase.

jbytheway commented:
OK, I've removed the item density fix and rebased on master.

Based on previous experience on this PR, the CI is going to take a long time to run (presumably because all the cached data is useless here).

akrieger commented on tests/list_test.cpp:
```diff
@@ -141,6 +140,7 @@ TEST_CASE( "list basics", "[list]" )
         SECTION( "reverse iterator count/access" ) {
```
Either this is lying or the loop below is wrong.

jbytheway replied:
Yeah; this is a copy/paste error.  Will fix.

akrieger commented on src/tgz_archiver.h:
```diff
@@ -2,6 +2,7 @@
 #define CATA_SRC_MAP_TGZ_ARCHIVER_H
 #include <array>
 #include <chrono>
+#include <filesystem>
```
Use our filesystem header filesystem.h or <ghc/fs_std_fwd.h>. Although given it already resolves fs::file_time_type, one or both are already transitively provided.

jbytheway replied:
Weird.  I don't know how this ever worked on C++14, but I will fix.

andrei8l replied:
It was already included through `"filesystem.h"` -> `<ghc/fs_std_fwd.hpp>` -> `<filesystem>`

https://github.com/CleverRaven/Cataclysm-DDA/blob/8286f4d7d8f43e1d3b79125ceb2270facf9e0caf/src/tgz_archiver.h#L8

akrieger commented on src/harvest.cpp:
```diff
@@ -43,20 +43,23 @@ bool string_id<harvest_drop_type>::is_valid() const
 {
-    return SNIPPET.random_from_category( succeeded ? msg_dissect_success : msg_dissect_fail ).value_or(
-               translation() );
+    // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDeleteLeaks)
```
?

jbytheway replied:
Something about the implementation of `unique_ptr` has changed in C++17 which causes what I believe to be a clang-tidy false-positive here.  I tried some tweaks to fix it, but nothing worked, so in the end I chose to suppress it.

Part of me is a little worried that it's not a false-positive and there really is a potential memory leak here, but the `value_ptr` implementation looks pretty solid to me...

jbytheway replied:
In case you're curious to analyse it yourself, here's the error message we get without the suppression:
```
.../Cataclysm-DDA/src/harvest.cpp:46:5: error: Potential leak of memory pointed to by field '_M_head_impl' [clang-analyzer-cplusplus.NewDeleteLeaks,-warnings-as-errors]
    return SNIPPET.random_from_category(
    ^
.../Cataclysm-DDA/src/harvest.cpp:47:16: note: Assuming 'succeeded' is false
               succeeded ? msg_fielddress_success : msg_fielddress_fail ).value_or( translation() );
               ^
.../Cataclysm-DDA/src/harvest.cpp:47:16: note: '?' condition is false
.../Cataclysm-DDA/src/harvest.cpp:46:12: note: Calling 'optional::value_or'
    return SNIPPET.random_from_category(
           ^
.../Cataclysm-DDA/src/optional.h:146:20: note: Assuming field 'full' is true
            return full ? get() : static_cast<T>( other );
                   ^
.../Cataclysm-DDA/src/optional.h:146:20: note: '?' condition is true
.../Cataclysm-DDA/src/optional.h:146:27: note: Calling implicit copy constructor for 'translation'
            return full ? get() : static_cast<T>( other );
                          ^
.../Cataclysm-DDA/src/optional.h:146:27: note: Calling copy constructor for 'value_ptr<std::basic_string<char>>'
.../Cataclysm-DDA/src/value_ptr.h:27:33: note: '?' condition is true
            std::unique_ptr<T>( other ? new T( *other ) : nullptr ) {}
                                ^
.../Cataclysm-DDA/src/value_ptr.h:27:41: note: Memory is allocated
            std::unique_ptr<T>( other ? new T( *other ) : nullptr ) {}
                                        ^
.../Cataclysm-DDA/src/optional.h:146:27: note: Returning from copy constructor for 'value_ptr<std::basic_string<char>>'
            return full ? get() : static_cast<T>( other );
                          ^
.../Cataclysm-DDA/src/optional.h:146:27: note: Returning from copy constructor for 'translation'
.../Cataclysm-DDA/src/harvest.cpp:46:12: note: Returned allocated memory
    return SNIPPET.random_from_category(
           ^
.../Cataclysm-DDA/src/harvest.cpp:46:5: note: Potential leak of memory pointed to by field '_M_head_impl'
    return SNIPPET.random_from_category(
    ^
```

BrettDong replied:
Looks similar to the type of false positive in https://bugs.llvm.org/show_bug.cgi?id=38176. Switching to `libc++` may help.

jbytheway replied:
It might help, but I'd prefer not to do that, especially not in this PR.  I'd like to keep what clang-tidy is analyzing as close as possible to what we build in the rest of CI.  And I don't think we use libc++ except on Mac.

Longer term, if we add the clang-tidy check that forbids any use of raw new or delete, then we can probably just disable this check.

akrieger replied:
@jbytheway if you resolve this then someone can merge this ;)

akrieger reviewed (approved):
A lot of this lint shouldve been caught pre-17, no?

jbytheway committed:
Fix copy/paste error in test name

jbytheway committed:
Don't include <filesystem>

This header can't be relied upon to exist.

jbytheway commented:
I think this is good to go, so I've removed the WIP label.
